packages/myst-transforms/src/indices.ts

```diff
@@ -0,0 +1,295 @@
+import type { Plugin } from 'unified';
+import type { GenericNode, GenericParent } from 'myst-common';
+import { copyNode, createId, fileWarn, normalizeLabel, RuleId } from 'myst-common';
+import { selectAll } from 'unist-util-select';
+import type { IReferenceStateResolver, ReferenceState, Target } from './enumerate.js';
+import type { VFile } from 'vfile';
+
+const ALPHABET_UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
+
+/**
+ * Ensure all nodes with index entries have label/identifier/html_id
+ */
+export function indexIdentifierTransform(mdast: GenericParent) {
+  const indexEntryNodes = selectAll('[indexEntries]', mdast) as GenericNode[];
+  indexEntryNodes.forEach((indexEntryNode) => {
+    if (!indexEntryNode.label) {
+      indexEntryNode.label = `index-${createId()}`;
+    }
+    if (!indexEntryNode.identifier) {
+      const { identifier, html_id } = normalizeLabel(indexEntryNode.label) ?? {};
+      indexEntryNode.identifier = identifier;
+      indexEntryNode.html_id = html_id;
+    }
+  });
+}
+
+export const indexIdentifierPlugin: Plugin<[], GenericParent, GenericParent> = () => (tree) => {
+  indexIdentifierTransform(tree);
+};
+
+type IndexTargetInfo = {
+  node: GenericNode;
+  emphasis?: boolean;
+};
+
+type IndexSeeInfo = {
+  subEntry: string;
+  emphasis?: boolean;
+};
+
+/**
+ * Take a list of index targets collected from a project and build lookup structure
+ *
+ * This sorts by starting letter, then top-level entry name, then sub-entry name
+ */
+function organizeTargetEntries(indexTargets: Target[], vfile: VFile) {
+  const entryDict: Record<
+    string, // Letter or Other
+    Record<
+      string, // Entry name
+      {
+        nodes: IndexTargetInfo[];
+        see: IndexSeeInfo[];
+        seeAlso: IndexSeeInfo[];
+        subEntries: Record<string, IndexTargetInfo[]>;
+      }
+    >
+  > = {};
+  indexTargets.forEach(({ node }) => {
+    node.indexEntries?.forEach(({ entry, subEntry, emphasis, see, seeAlso }) => {
+      const letter = ALPHABET_UPPER.includes(entry[0]?.toUpperCase())
```

Comment:
This means we don't support grouping by number (only letters). Is that intentional?

Reply:
This is intentional, only to match legacy jupyter-book behavior. Entries that start with numbers appear under `Symbols`:
![image](https://github.com/user-attachments/assets/eb39bc75-2d89-4623-8ba1-3228158d0970)

Could be a nice area to improve with user-facing directive options? But for now, I think we just keep this behavior.